src/process/path.h

```diff
@@ -154,6 +199,25 @@ static __always_inline int write_path(struct pt_regs *ctx, cached_path_t *cached
         if (bpf_probe_read(&offset, sizeof(offset), dentry + name) < 0)
             goto NameError;
 
+        // Path filtering, only if the filter is active
+        if (cached_path->filter_state >= 0) {
+            val = NULL;
+            // We need to reset at least the whole string storage array each time we use the key
+            bpf_probe_read(&key, sizeof(filter_key_t), &blank_key);
+            key.current_state = cached_path->filter_state;
+            if (bpf_probe_read_str(&key.path_segment, MAX_PATH_SEG, offset) > 0) {
+                val = (filter_value_t *)bpf_map_lookup_elem(&filter_transitions, &key);
+            }
+            // One retry with "*" path element if not found
+            if (val == NULL) {
+                // The retry_key always uses the same path segment string so we don't need to reset
+                retry_key.current_state = cached_path->filter_state;
+                val = (filter_value_t *)bpf_map_lookup_elem(&filter_transitions, &retry_key);
+            } 
+            // Terminate the filter lookup if there was no value at either key
+            cached_path->filter_state = val ? val->next_state : -1;
+        }
+
```

Comment:
This code is re-used in the execve* kretprobes which were already pretty close to its max instructions and now they are going over  (even though it will never hit that if but the verifier doesn't know that). The easiest "fix" would be to reduce the number of segments we can handle per tail (`MAX_PATH_SEGMENTS_NOTAIL`) which is currently at 25 but it'd be cool if we could find different way

Reply:
Since path.h is included in both programs via the preprocessor, I have used a preprocessor variable and `#if` blocks to allow programs to choose whether to include the filter stuff or not. It's a little ugly, but ensures the exec programs don't have to deal with the overhead.